examples/diff.c

```diff
@@ -67,13 +75,70 @@ static int color_printer(
 	const git_diff_delta*, const git_diff_hunk*, const git_diff_line*, void*);
 static void diff_print_stats(git_diff *diff, struct opts *o);
 
+/*
+ * Read a file path
+ *
+ * path: a file path (char*)
+ *
+ * returns char* with the contents of the file (must be freed) or NULL if
+ * file does not exist or memory allocation failed
+ */
+char* read_file(const char* path) {
+    FILE* file;
+	 char* result;
+    size_t pos = 0;
+    size_t capacity = READ_SIZE;
+    char ch;
+
+    file = fopen(path, "r");
+
+    if (!file) {
+        return NULL;
+	 }
+
+    result = malloc(sizeof(char) * READ_SIZE + 1);
+
+    if (result == NULL) {
+        return NULL;
+	 }
+
+    while ((ch = getc(file)) != EOF) {
+        result[pos++] = ch;
+
+        if(pos >= capacity) {
+            capacity += READ_SIZE;
+
+            result = realloc(result, sizeof(char) * capacity + 1);
+
+            if(result == NULL) {
+                return NULL;
+				}
+        }
+    }
+
+    fclose(file);
+
+    result = realloc(result, sizeof(char) * pos);
```

Comment:
This extra reallocation might not be worth it. You gain a maximum of `READ_SIZE-1` bytes, which is just a quarter of a kb with your default value. Unless `READ_SIZE` is much greater than the size of a memory page or you allocate some _small_ objects afterwards, you will not gain any memory back. It doesn't matter for a short-lived process anyway.